src/librbd/ManagedLock.cc

```diff
@@ -0,0 +1,661 @@
+}
+
+template <typename I>
+void ManagedLock<I>::handle_reacquire_lock(int r) {
+  ldout(m_cct, 10) << ": r=" << r << dendl;
+
+  Mutex::Locker locker(m_lock);
+  assert(m_state == STATE_REACQUIRING);
+
+  if (r < 0) {
+    if (r == -EOPNOTSUPP) {
+      ldout(m_cct, 10) << ": updating lock is not supported" << dendl;
+    } else {
+      lderr(m_cct) << ": failed to update lock cookie: " << cpp_strerror(r)
+                   << dendl;
+    }
+
+    if (!is_shutdown_locked()) {
+      // queue a release and re-acquire of the lock since cookie cannot
+      // be updated on older OSDs
+      execute_action(ACTION_RELEASE_LOCK, nullptr);
+
+      assert(!m_actions_contexts.empty());
+      ActionContexts &action_contexts(m_actions_contexts.front());
+
+      // reacquire completes when the request lock completes
+      Contexts contexts;
+      std::swap(contexts, action_contexts.second);
+      if (contexts.empty()) {
+        execute_action(ACTION_ACQUIRE_LOCK, nullptr);
+      } else {
+        for (auto ctx : contexts) {
+          ctx = new FunctionContext([ctx, r](int acquire_ret_val) {
+              if (acquire_ret_val >= 0) {
+                acquire_ret_val = r;
+              }
+              ctx->complete(acquire_ret_val);
+            });
+          execute_action(ACTION_ACQUIRE_LOCK, ctx);
+        }
+      }
+    }
+  } else {
+    m_cookie = m_new_cookie;
+  }
+
+  complete_active_action(STATE_LOCKED, r);
+}
+
+template <typename I>
+void ManagedLock<I>::send_release_lock() {
+  assert(m_lock.is_locked());
+  if (m_state == STATE_UNLOCKED) {
+    complete_active_action(STATE_UNLOCKED, 0);
+    return;
+  }
+
+  ldout(m_cct, 10) << dendl;
+  m_state = STATE_RELEASING;
+
+  pre_release_lock_handler(false, util::create_context_callback<
+      ManagedLock<I>, &ManagedLock<I>::handle_pre_release_lock>(this));
+}
+
+template <typename I>
+void ManagedLock<I>::handle_pre_release_lock(int r) {
+  ldout(m_cct, 10) << ": r=" << r << dendl;
+
+  using managed_lock::ReleaseRequest;
+  ReleaseRequest<I>* req = ReleaseRequest<I>::create(m_ioctx, m_watcher,
+      m_work_queue, m_oid, m_cookie,
+      util::create_context_callback<
+        ManagedLock<I>, &ManagedLock<I>::handle_release_lock>(this));
+  m_work_queue->queue(new C_SendLockRequest<ReleaseRequest<I>>(req), 0);
+}
+
+template <typename I>
+void ManagedLock<I>::handle_release_lock(int r) {
+  ldout(m_cct, 10) << ": r=" << r << dendl;
+
+  Mutex::Locker locker(m_lock);
+  assert(m_state == STATE_RELEASING);
+
+  if (r >= 0) {
+    m_cookie = "";
+  }
+
+  post_release_lock_handler(false, r, util::create_context_callback<
+      ManagedLock<I>, &ManagedLock<I>::handle_post_release_lock>(this));
+}
+
+template <typename I>
+void ManagedLock<I>::handle_post_release_lock(int r) {
+  ldout(m_cct, 10) << ": r=" << r << dendl;
+
+  assert(m_lock.is_locked());
+  complete_active_action(r < 0 ? STATE_LOCKED : STATE_UNLOCKED, r);
```

Comment:
If the failure occurred at the derived class handler -- the lock might have been released but you revert to the LOCKED state.